### PR TITLE
Extended QPT to athena-dynamodb

### PR DESCRIPTION
**Pull request by AbdulR3hman**
*Description of changes:*
Will allow customer to execute QPT on DynamoDB using PartiQL; Only SELECT statements are supported.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBMetadataHandler.java [hunk 1]
```diff
@@ -230,6 +250,27 @@ public ListTablesResponse doListTables(BlockAllocator allocator, ListTablesReque
...
+        if (!request.isQueryPassthrough()) {
+            throw new IllegalArgumentException("No Query passed through [{}]" + request);
+        }
+
+        queryPassthrough.verify(request.getQueryPassthroughArguments());
+        String partiQLStatement = request.getQueryPassthroughArguments().get(DDBQueryPassthrough.QUERY);
+        ExecuteStatementRequest executeStatementRequest =
+                ExecuteStatementRequest.builder()
+                        .statement(partiQLStatement)
+                        .limit(4)
```
Comment (chngpe) [inline athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBMetadataHandler.java:265]: use `SCHEMA_INFERENCE_NUM_RECORDS` from DDBTableUtils instead?

File: athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/qpt/DDBQueryPassthrough.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if (!upperCaseStatement.startsWith("SELECT")) {
+            throw new UnsupportedOperationException("Statement does not start with SELECT.");
+        }
+
+        // List of disallowed keywords
+        Set<String> disallowedKeywords = ImmutableSet.of("INSERT", "UPDATE", "DELETE", "CREATE", "DROP", "ALTER");
+
+        // Check if the statement contains any disallowed keywords
+        for (String keyword : disallowedKeywords) {
+            if (upperCaseStatement.contains(" " + keyword + " ") || upperCaseStatement.startsWith(keyword + " ")) {
```
Comment (chngpe) [inline athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/qpt/DDBQueryPassthrough.java:88]: can we have just contains here? or are we assuming there will be column name as `insert`  here? Those are reserved word in DDB, so i think should be okay. https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ReservedWords.html

File: athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTableUtils.java [hunk 3]
```diff
@@ -187,6 +175,24 @@ public static Schema peekTableForSchema(String tableName, ThrottlingInvoker invo
         return schemaBuilder.build();
     }
 
+    public static SchemaBuilder buildSchemaFromItems(List<Map<String, AttributeValue>> items)
+    {
+        SchemaBuilder schemaBuilder = new SchemaBuilder();
+        Set<String> discoveredColumns = new HashSet<>();
+        for (Map<String, AttributeValue> item : items) {
+            for (Map.Entry<String, AttributeValue> column : item.entrySet()) {
+                if (!discoveredColumns.contains(column.getKey())) {
```
Comment (chngpe) [inline athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTableUtils.java:184]: can we improve it to not have this many layer?

File: athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBMetadataHandler.java (hunk 1 above)
Comment (AbdulR3hman) [reply athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBMetadataHandler.java:265]: good catch; yes will do.

File: athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/qpt/DDBQueryPassthrough.java (hunk 2 above)
Comment (AbdulR3hman) [reply athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/qpt/DDBQueryPassthrough.java:88]: will make the fix. and yes I don't think insert is allowed as its reserved keyword.

File: athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTableUtils.java (hunk 3 above)
Comment (AbdulR3hman) [reply athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTableUtils.java:184]: Yes, I'll refactor it.

**Review by chngpe — APPROVED** [review]

**Pull request merged**